src/software/backend/simulation/simulator_robot_test.cpp

```diff
@@ -3,39 +3,688 @@
+{
+    Robot robot(0, Point(0, 0), Vector(0, 0), Angle::zero(), AngularVelocity::zero(),
+                Timestamp::fromSeconds(0));
+    Ball ball(Point(0.15, 0), Vector(-0.25, 0), Timestamp::fromSeconds(0));
+    std::vector<Point> enemy_robot_positions = {Point(0.4, 0), Point(1.9, 0)};
+    auto [world, simulator_robot, simulator_ball] =
+        createWorldWithEnemyRobots(robot, ball, enemy_robot_positions);
+
+    EXPECT_LT((simulator_ball->velocity() - Vector(-0.25, 0)).length(), 0.001);
+
+    simulator_robot->enableAutochip(2);
+
+    // Simulate for 1.5 seconds
+    for (unsigned int i = 0; i < 90; i++)
+    {
+        world->stepSimulation(Duration::fromSeconds(1.0 / 60.0));
+    }
+
+    EXPECT_GT(simulator_ball->velocity().x(), 1);
+    EXPECT_GT(simulator_ball->position().x(), 2.0);
+}
+
+TEST_F(SimulatorRobotTest,
+       test_robot_chips_ball_with_autochip_enabled_and_ball_lands_on_obstacle)
+{
+    Robot robot(0, Point(0, 0), Vector(0, 0), Angle::zero(), AngularVelocity::zero(),
+                Timestamp::fromSeconds(0));
+    Ball ball(Point(0.15, 0), Vector(-0.25, 0), Timestamp::fromSeconds(0));
+    // Add an enemy robot right where the ball will land
+    std::vector<Point> enemy_robot_positions = {Point(2.0, 0)};
+    auto [world, simulator_robot, simulator_ball] =
+        createWorldWithEnemyRobots(robot, ball, enemy_robot_positions);
+
+    EXPECT_LT((simulator_ball->velocity() - Vector(-0.25, 0)).length(), 0.001);
+
+    simulator_robot->enableAutochip(2);
+
+    // Simulate for 1.5 seconds
+    for (unsigned int i = 0; i < 90; i++)
+    {
+        world->stepSimulation(Duration::fromSeconds(1.0 / 60.0));
+    }
+
+    // Check the ball's velocity is very roughly within a reasonable range. We don't care
+    // exactly what it is, just that it hasn't changed an impossible amount by
+    // getting caught inside the robot it landed in and then squeezed out
+    EXPECT_GT(simulator_ball->velocity().x(), 1);
+    EXPECT_LT(simulator_ball->velocity().x(), 3);
+    EXPECT_GT(simulator_ball->position().x(), 2.0);
+}
+
+TEST_F(SimulatorRobotTest, test_ball_does_not_bounce_off_front_of_robot_when_dribbler_on)
+{
+    Robot robot(0, Point(0, 0), Vector(0, 0), Angle::zero(), AngularVelocity::zero(),
+                Timestamp::fromSeconds(0));
+    Ball ball(Point(0.25, 0), Vector(-3, 0), Timestamp::fromSeconds(0));
+    auto [world, simulator_robot, simulator_ball] = createWorld(robot, ball);
+
+    EXPECT_LT((simulator_ball->velocity() - Vector(-3, 0)).length(), 0.001);
+
+    // We use an arbitrarily large number here for speed
+    simulator_robot->setDribblerSpeed(10000);
+
+    // Simulate for 2 seconds
+    for (unsigned int i = 0; i < 120; i++)
+    {
+        world->stepSimulation(Duration::fromSeconds(1.0 / 60.0));
+    }
+
+    // Make sure we haven't caused the robot to move significantly by dribbling
+    EXPECT_LT((simulator_robot->position() - Point(0, 0)).length(), 0.003);
```

Comment:
Perhaps we could apply equal and opposite force to the robot body to prevent the robot from moving at all?

Reply:
Possibly, but that then only works if the ball _is_ touching the robot/chicker most of the time, which we currently have setup so that it doesn't quit do that, and rather applied 0 force if the ball is super close so we don't apply a net force.

I've determined that (at least in this test) no matter how long the simulation runs the robot only moves a tiny fixed amount, so I think it's fine to leave for now.